### PR TITLE
[81] Redo employment gap detection

### PR DESCRIPTION
**Pull request by elliotcm**
This is a fresh version of the logic with comprehensive unit tests.

This is hard to screenshot given all the variations.

### CODE REVIEW

**Comment by elliotcm** [comment]
[Investigate/Refactor gap in employment](https://trello.com/c/8MBVRSv4/256-investigate-refactor-gap-in-employment)

File: app/services/jobseekers/job_applications/employment_gap_finder.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  private
+
+  attr_reader :job_application, :today
+
+  def gap_to_today_is_less_than_threshold?(employment, threshold)
+    employment.ended_on + threshold >= today
+  end
+
+  def overlapping_employment?(employment)
+    job_application.employments.any? do |other|
```
Comment (josephhull676) [inline app/services/jobseekers/job_applications/employment_gap_finder.rb:30]: Could you do something like `.excluding(employment)` here?

Could save on the extra line `next if employment == other`

File: app/services/jobseekers/job_applications/employment_gap_finder.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+  def next_employment_started_within_threshold?(employment, threshold)
+    return unless (next_start = next_employment_start(employment))
+
+    next_start <= employment.ended_on + threshold
+  end
+
+  def next_employment_start(employment)
+    job_application.employments
+      .reject { |other| other == employment }
```
Comment (josephhull676) [inline app/services/jobseekers/job_applications/employment_gap_finder.rb:49]: And perhaps here? 🤷‍♂️

**Review by josephhull676 — APPROVED** [review]

File: app/services/jobseekers/job_applications/employment_gap_finder.rb (hunk 1 above)
Comment (elliotcm) [reply app/services/jobseekers/job_applications/employment_gap_finder.rb:30]: @josephhull676 This is not a bad idea, but having tried it just now the problem is that only works if all the records are saved to the database, whereas the current implementation will work on any array of employments.

**Comment by bencmitchell** [comment]
This looks good to me 👍

File: app/services/jobseekers/job_applications/employment_gap_finder.rb (hunk 1 above)
Comment (josephhull676) [reply app/services/jobseekers/job_applications/employment_gap_finder.rb:30]: Ahh ok 👍

**Pull request merged**